File: 20221102-removeParing-java/app/src/main/java/RemoveParing.java
// 이해
// 주어진 것 : 문자열
// 조건: 앞에서부터 주어진 문자열에서 같은 알파벳이 붙어 있는 2개를 우선적으로 찾는다. -> 제거한다.
// 위 과정을 하고 남은 것을 다시 또 반복한다.
// 더 이상 제거할 수 없을때까지 반복한다.
// 계획
// 어떻게 반복한 것을 제거하지? -> for문? stream filter?
// 어떻게 연속된 문자가 있다는 것을 검사하지? -> 앞의 원소가 뒤와 같다?
// 어떻게 제거하지? 제자리 인덱스 2번 제거

import java.util.*;

public class RemoveParing {
  public int solution(String s) {

    Stack<Character> stack = new Stack();

    for (char element : s.toCharArray()) {
      if (!stack.isEmpty() && stack.peek().equals(element)) {
        stack.pop();
      } else {
        stack.push(element);
      }
    }
    return stack.isEmpty() ?  1 :  0;
  }
}

File: 20221103-removeParing-javaScript/RemoveParing.js
// 이해
// 1.  노란격자를 브라운이 1줄로 감싸고 있음
// 2. 노란격자 수와 브라운 격자 수가 주어짐
// 3. 노란격자는 항상 가로롤 넓음
// 풀이
// 1. 브라운 갯수 = (노란색의 세로길이 + 2)* 2 + (노란색의 가로길이) * 2
// 2. 노랑이 가로 + 2 + 가로길이 노랑이 세로 + 2

export default function solution(s) {
  // if (s.length % 2 !== 0) return 0;
  const stack = [];

  [...s].forEach((element) => {
    console.log('s', [...s]);

    if (stack[stack.length - 1] === element) {
      stack.pop();
      console.log('stackPop', stack);
      return;
    }
    if (stack[stack.length - 1] !== element) {
      stack.push(element);
      console.log('stackPush', stack);
    }
  });

  // for (let i = 0; i < s.length; i++) {
  //   const b = s.charAt(i);
  //   if (stack[stack.length - 1] === b) {
  //     stack.pop();
  //     console.log('stackPop', stack);
  //   } else {
  //     stack.push(b);
  //     console.log('stackPush', stack);
  //   }
  // }
  // console.log('finalStack1',stack);
  return stack.length > 0 ? 0 : 1;
}
